Circuit Board Vending Machine

Cherman0 opened the pull request:
<!-- The text between the arrows are comments - they won't be visible on your PR. -->
<!-- To label this PR, add the label(s) without the prefixes surrounded by brackets anywhere, for example: [LABEL] -->
<!-- PRs should at least have one area (A-) label and at least one category (C-) label -->

## About the PR <!-- Describe the Pull Request here. What does it change? What other things could this impact? -->

![image](https://github.com/user-attachments/assets/f1a9ae2b-81e0-406c-a38d-40d6fddd1870)
Adds a vending machine that holds the circuit boards needed to create specialized consoles. Does not add them to any maps. Most boards cost 50c, though some of the hidden ones are more or less expensive.

## Why's this needed? <!-- Describe why you think this should be added to the game. -->

Allows access to every circuit board, not just the handful that spawn in the piles in tech storage. It would be impractical to scale the pile of circuit boards as the number of boards in the game increases. Vending machines support a scrollbar and ctrl+f searching, making them better suited to large lists of items.

JORJ949 commented on code/modules/vending/vending.dm:
```diff
@@ -1707,6 +1707,71 @@ ABSTRACT_TYPE(/obj/machinery/vending/cola)
+		product_list += new/datum/data/vending_product(/obj/item/circuitboard/secure_data, amount=2, cost=PAY_UNTRAINED/3)
+		product_list += new/datum/data/vending_product(/obj/item/circuitboard/atmospherealerts, amount=2, cost=PAY_UNTRAINED/3)
+		product_list += new/datum/data/vending_product(/obj/item/circuitboard/atmospheresiphonswitch, amount=2, cost=PAY_UNTRAINED/3)
+		product_list += new/datum/data/vending_product(/obj/item/circuitboard/air_management, amount=2, cost=PAY_UNTRAINED/3)
```
2 entries for air_management

JORJ949 commented on code/modules/vending/vending.dm:
```diff
@@ -1707,6 +1707,71 @@ ABSTRACT_TYPE(/obj/machinery/vending/cola)
+		product_list += new/datum/data/vending_product(/obj/item/circuitboard/sea_elevator, amount=2, cost=PAY_UNTRAINED/3)
+		//Hidden circuits available by hacking:
+		product_list += new/datum/data/vending_product(/obj/item/circuitboard/clown_announcement, amount=1, hidden=TRUE, cost=PAY_DUMBCLOWN)
+		product_list += new/datum/data/vending_product(/obj/item/circuitboard/robotics, amount=1, hidden=TRUE, cost=PAY_EXECUTIVE)
```
If you want this used anywhere ever, it should not have the killswitch, that should take effort to steal.

soleilan replied:
Ditto, those are usually placed in computer core as a circuit board on all maps for a reason

JORJ949 commented on code/modules/vending/vending.dm:
```diff
@@ -1707,6 +1707,71 @@ ABSTRACT_TYPE(/obj/machinery/vending/cola)
+		product_list += new/datum/data/vending_product(/obj/item/circuitboard/security, amount=1, hidden=TRUE, cost=PAY_EXECUTIVE)
+		product_list += new/datum/data/vending_product(/obj/item/circuitboard/bank_data, amount=1, hidden=TRUE, cost=PAY_EXECUTIVE)
+		product_list += new/datum/data/vending_product(/obj/item/circuitboard/card, amount=1, hidden=TRUE, cost=PAY_DONTBUYIT)
+		//not present because because deprecated/nonfunctional to build:
```
If deprecated, why add at all?

JORJ949 commented on code/modules/vending/vending.dm:
```diff
@@ -1707,6 +1707,71 @@ ABSTRACT_TYPE(/obj/machinery/vending/cola)
+		product_list += new/datum/data/vending_product(/obj/item/circuitboard/qmorder, amount=2, cost=PAY_UNTRAINED/3)
+		product_list += new/datum/data/vending_product(/obj/item/circuitboard/qmsupply, amount=2, cost=PAY_UNTRAINED/3)
+		product_list += new/datum/data/vending_product(/obj/item/circuitboard/stockexchange, amount=2, cost=PAY_UNTRAINED/3)
+		product_list += new/datum/data/vending_product(/obj/item/circuitboard/transception, amount=2, cost=PAY_UNTRAINED/3)
```
Should have a check here for if its nadir so you dont end up with pointless transception controls

JORJ949 reviewed (commented):
Review mostly made assuming this is intended to be player-usable (otherwise why add it?)

soleilan commented on code/modules/vending/vending.dm:
```diff
@@ -1707,6 +1707,71 @@ ABSTRACT_TYPE(/obj/machinery/vending/cola)
+		product_list += new/datum/data/vending_product(/obj/item/circuitboard/mining_magnet, amount=2, cost=PAY_UNTRAINED/3)
+		product_list += new/datum/data/vending_product(/obj/item/circuitboard/telescope, amount=2, cost=PAY_UNTRAINED/3)
+		product_list += new/datum/data/vending_product(/obj/item/circuitboard/announcement, amount=2, cost=PAY_UNTRAINED/3)
+		product_list += new/datum/data/vending_product(/obj/item/circuitboard/siphon_control, amount=2, cost=PAY_UNTRAINED/3)
```
Include this in Nadir check

soleilan commented on code/modules/vending/vending.dm:
```diff
@@ -1707,6 +1707,71 @@ ABSTRACT_TYPE(/obj/machinery/vending/cola)
+		product_list += new/datum/data/vending_product(/obj/item/circuitboard/siphon_control, amount=2, cost=PAY_UNTRAINED/3)
+		product_list += new/datum/data/vending_product(/obj/item/circuitboard/chem_request, amount=2, cost=PAY_UNTRAINED/3)
+		product_list += new/datum/data/vending_product(/obj/item/circuitboard/chem_request_receiver, amount=2, cost=PAY_UNTRAINED/3)
+		product_list += new/datum/data/vending_product(/obj/item/circuitboard/sea_elevator, amount=2, cost=PAY_UNTRAINED/3)
```
Include this in Oshan/Nadir check

pgmzeta commented:
Can you please explain the rationale behind charging credits for these?

Cherman0 commented:
Mostly because some of the other tech storage vendors also do. In theory the price could be balancing but the funny free stuff mechcomp button exists, so its basically just fluff.

As for map checking, I'd rather it didn't. Map checks add extra random spots that need changed if maps are added or changed, so as long as it doesnt introduce broken functionality I'd rather just have items that arent useful.

Im not against removing robotics or id computer boards, though since you need a command ID to actually use them it doesnt seem like an issue to me. Since at that point you could just use the robotics console in the bridge or ai areas.

Cherman0 commented:
Removed the robot control board. Also fixed that the computer type of the clown announcement computer was a string instead of a typepath, which would prevent it from being used to make a console.

Cherman0 commented:
Using mapchecks to minimize the size of a list of items is not really necessary for vending machines, since they're much easier to navigate than the rightclick list of items on the ground. You could similarly argue that the pump control boards shouldn't be available on maps with no pumps, but there's no real reason to remove them and it just adds work later.

JORJ949 commented:
> Using mapchecks to minimize the size of a list of items is not really necessary for vending machines, since they're much easier to navigate than the rightclick list of items on the ground.
>

All the boards look the same unlike other vending machines which you can visibly see so you have to read, cutting down unnecessary bloat I think there isnt much reason to not do, pumps are on almost every map and solars you just do a NOT oshan or nadir check for

soleilan commented:
Yeah what they said, but also think about it like this. What if making a sea elevator computer throws mad runtime errors on maps without one? Are you going to go through all of these circuits and make sure they don't throw errors on maps that don't support them? It's just much easier to make sure they can't be built at all. Pumps are on every single map IIRC since they are part of Toxins but yeah I would add logic for solars, transception arrays, sea elevators, siphon control. Even if they don't throw errors, there's literally no reason for them to be buildable and confuse new players or invite people to find ways to make them throw errors IMO.

Cherman0 commented:
Eh, sure. It's added.

Cherman0 commented:
And for context, I haven't added it to maps as a matter of scope. I'd rather not fix map merge conflicts all day while the matter at hand is wether it's actually wanted in the first place at all.

soleilan reviewed (approved):
Can't speak on the topic if its wanted or about the balance, but code-wise it looks all good.

JORJ949 reviewed (approved):
Looks good to me, though I think it should be added to maps in the same PR tbh.
